javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1014,6 +1043,201 @@ object Formatter {
+      }
+    }
+  }
+
+  /** A decimal representation of a number.
+   *
+   *  An instance of this class represents the number whose absolute value is
+   *  `unscaledValue × 10^(-scale)`, and that is negative iff `negative` is
+   *  true.
+   *
+   *  The `unscaledValue` is stored as a String of decimal digits, i.e.,
+   *  characters in the range ['0', '9'], expressed in base 10. Leading 0's are
+   *  *not* valid.
+   *
+   *  As an exception, a zero value is represented by an `unscaledValue` of
+   *  `"0"`. The scale of zero value is always 0.
+   *
+   *  `Decimal` is similar to `BigDecimal`, with some differences:
+   *
+   *  - `Decimal` distinguishes +0 from -0.
+   *  - The unscaled value of `Decimal` is stored in base 10.
+   *
+   *  The methods it exposes have the same meaning as for BigDecimal, with the
+   *  only rounding mode being HALF_UP.
+   */
+  private final class Decimal(val negative: Boolean, val unscaledValue: String,
+      val scale: Int) {
+
+    def isZero: Boolean = unscaledValue == "0"
+
+    /** The number of digits in the unscaled value.
+     *
+     *  The precision of a zero value is 1.
+     */
+    def precision: Int = unscaledValue.length()
+
+    /** Rounds the number so that it has at most the given precision, i.e., at
+     *  most the given number of digits in its `unscaledValue`.
+     *
+     *  The given `precision` must be greater than 0.
+     */
+    def round(precision: Int): Decimal = {
+      if (precision <= 0)
+        throw new AssertionError("Decimal.round() called with non-positive precision")
+
+      if (precision >= unscaledValue.length()) {
+        // no rounding needed (this branch is always taken for zero values)
+        this
+      } else {
+        val diff = unscaledValue.length() - precision
+        val rounded = roundDigitsAtPos(unscaledValue, precision)
+        val roundedLen = rounded.length()
+
+        if (roundedLen > precision) {
+          // the carry expanded all the way to a new digit
+          if (rounded.charAt(roundedLen - 1) != '0')
+            throw new AssertionError(s"$this.round($precision) caused an invalid rounding")
+          val amendedDigits = rounded.substring(0, roundedLen - 1)
+          val newScale = scale - diff - 1
+          new Decimal(negative, amendedDigits, newScale)
+        } else {
+          val newScale = scale - diff
+          new Decimal(negative, rounded, newScale)
+        }
+      }
+    }
+
+    /** Returns a new Decimal instance with the same value, possibly rounded,
+     *  with the given scale.
+     *
+     *  If this is a zero value, the same value is returned (a zero value must
+     *  always have a 0 scale). Rounding may also cause the result to be a zero
+     *  value, in which case its scale must be 0 as well. Otherwise, the result
+     *  is non-zero and is guaranteed to have exactly the given new scale.
+     */
+    def setScale(newScale: Int): Decimal = {
```

Comment:
Have you considered something like this?

```scala
val newPrecision = unscaledValue.length() - scale + newScale

if (newPrecision <= 0) {
  Decimal.zero(negative)
} else {
  val rounded = round(newPrecision)
  assert(rounded.scale <= newScale)
  if (rounded.scale < newScale)
    new Decimal(negative, rounded.unscaledValue + strOfZeros(newScale - rounded.scale), newScale)
  else
    rounded
}
```

Reply:
Hum, I have not. I'll play around with that as well.

Reply:
This is the simplest I could make `setScale` to be if implemented on top of `round`:
```scala
    def setScale(newScale: Int): Decimal = {
      val targetPrecision = precision + newScale - scale

      if (targetPrecision < 0) {
        Decimal.zero(negative)
      } else if (targetPrecision == 0) {
        // Weird case: handle 0.3 -> 0 and 0.7 -> 1 (round cannot do this)
        if (unscaledValue.charAt(0) < '5')
          Decimal.zero(negative)
        else
          new Decimal(negative, "1", newScale)
      } else {
        val rounded = this.round(targetPrecision)
        // Assert: rounded.isZero || rounded.scale <= newScale
        if (rounded.isZero || rounded.scale == newScale)
          rounded
        else
          new Decimal(negative, rounded.unscaledValue + strOfZeros(newScale - rounded.scale), newScale)
      }
    }
```
I am not sure at all that this is simpler than the previous version.

Reply:
Hmmm... It is definitely not. Would it get simpler if we made an internal `roundAtPos` that allowed at `0`? (and `round` could basically forward to that one).

Reply:
Well, that required quite a bit of refactoring, but it worked. I pushed a new commit with a simplified implementation of `round` and `setScale` in terms of a `roundAtPos`.

In addition, I renamed the `precision` parameter of `scientificNotation` and `decimalNotation` to make them clearer, so that we don't overload `precision`.

Does it help?

Reply:
Yes. TY :)